resources/templates/md/posts/2016-04-04-state.md

```diff
@@ -0,0 +1,370 @@
+
+In the bag there are 30 marshmallows. So we will define `marshmallows-in-the-bag` with a value `30`.
+
+```clojure
+(def marshmallows-in-the-bag 30)
+```
+
+Now we will define a function `marshmallows-left`. This function takes one argument `marshmallows-eaten`. 
+In a let block we will define the `new-amount-of-marshmallows` and the value of this Var is the difference between `marshmallows-in-the-bag`
+and `marshmallows-eaten`. 
+As a last step in this function is the println `new-amount-of-marshmallows`.
+
+```clojure
+(defn marshmallows-left [marshmallows-eaten]
+  (let [new-amount-of-marshmallows (- marshmallows-in-the-bag marshmallows-eaten)]
+     (println (str "Marshmallows in the bag: " new-amount-of-marshmallows))))
+```
+
+Calling our function `marshmallows-left` with 10 (because Greg ate 10 pieces of marshmallow)
+
+```Clojure
+(marshmallows-left 10)
+
+;;=> Marshmallows in the bag: 20
+```
+We found out that in the bag 20 marshmallows left.
+
+### Dynamic Scope
+
+In difference to `lexical scope`, the dynamic scope does not depend on code block, but on the runtime call stack.
+
+The best way to understand the difference is to look at the code below.
+
+```Clojure
+(def ^:dynamic *marshmallows-in-the-bag* 30)
+
+(defn marshmallows-left []
+  (println (str "Marshmallows in the bag: " *marshmallows-in-the-bag*)))
+
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (binding [*marshmallows-in-the-bag* (- *marshmallows-in-the-bag* marshmallows-eaten)]
+    (marshmallows-left)))
+```
+
+We created a function `marshmallows-left`. This function prints the value of our Var `marshmallows-in-the-bag`. 
+And we have another function `number-of-marshmallows-left` with one argument `marshmallows-eaten`. 
+In this function we bind the new value for `marshmallows-in-the-bag` (difference between `marshmallows-in-the-bag` and `marshmallows-eaten`).
+
+Now we call `number-of-marshmallows-left` with 10 as an argument (as we know, Greg ate 10 marshmallows). 
+
+```Clojure
+(number-of-marshmallows-left 10)
+
+;;=> Marshmallows: 20
+```
+
+To show the difference we'll do the same using `let block` (`lexical scope`) instead of `binding` (`dynamic scope`). 
+
+```Clojure
+(def ^:dynamic *marshmallows-in-the-bag* 30)
+
+(defn marshmallows-left []
+  (println (str "Marshmallows: " *marshmallows-in-the-bag*)))
+
+(defn number-of-marshmallows-left [marshmallows-eaten]
+  (let [*marshmallows-in-the-bag* (- *marshmallows-in-the-bag* marshmallows-eaten)]
+    (marshmallows-left)))
+```
+
+And if we call `numbers-of-marshmallows-left` with `10` as an argument, the result will be different as before.
+
+```Clojure
+(number-of-marshmallows-left 10)
+
+;;=> Marshmallows: 30
+```
+
+That's because we defined `marshmallows-in-the-bag` in the `let` code block and as we know the new value of the `marshmallows-in-the-bag` 
+is valid only within it's lexical scope. So for the `marshmallows-left` function its the value of the `marshmallows-in-the bag` still `30`. 
```

Comment:
is valid only within _its_ lexical scope. 
...
So for the `marshmallows-left` function _it's_ the value